diff --git a/.gitlab-ci.yml b/.gitlab-ci.yml
index 61fbfafba..cf797105b 100644
--- a/.gitlab-ci.yml
+++ b/.gitlab-ci.yml
@@ -77,12 +77,13 @@ include:
 .ubuntu:
   variables:
     # Update this tag when you want to trigger a rebuild
-    FDO_DISTRIBUTION_TAG: '2022-01-27.0'
+    FDO_DISTRIBUTION_TAG: '2022-11-23.1'
     FDO_DISTRIBUTION_VERSION: '20.04'
     FDO_DISTRIBUTION_PACKAGES: >-
       debhelper-compat
       findutils
       git
+      libapparmor-dev
       libasound2-dev
       libavcodec-dev
       libavfilter-dev
@@ -93,6 +94,7 @@ include:
       libgstreamer-plugins-base1.0-dev
       libsbc-dev
       libsdl2-dev
+      libsnapd-glib-dev
       libudev-dev
       libva-dev
       libv4l-dev
@@ -227,7 +229,7 @@ build_on_ubuntu:
     - .build
   stage: build
   variables:
-    MESON_OPTIONS: "-Dsession-managers=[]"
+    MESON_OPTIONS: "-Dsession-managers=[] -Dsnap=enabled"
 
 .build_on_fedora:
   extends:
@@ -254,6 +256,7 @@ build_on_fedora:
         -Dsdl2=enabled
         -Dsndfile=enabled
         -Dsession-managers=[]
+        -Dsnap=disabled
   artifacts:
     name: pipewire-$CI_COMMIT_SHA
     when: always
@@ -269,7 +272,7 @@ build_on_alpine:
     - .build
   stage: build
   variables:
-    MESON_OPTIONS: "-Dsession-managers=[]"
+    MESON_OPTIONS: "-Dsession-managers=[] -Dsnap=disabled"
 
 # build with all auto() options enabled
 build_all:
@@ -286,6 +289,7 @@ build_all:
         -Droc=disabled
         -Dlibcamera=disabled
         -Dsession-managers=[]
+        -Dsnap=disabled
   parallel:
     matrix:
       - CC: [gcc, clang]
@@ -295,7 +299,7 @@ build_with_no_commandline_options:
   extends:
     - .build_on_fedora
   variables:
-    MESON_OPTIONS: "-Dsession-managers=[]"
+    MESON_OPTIONS: "-Dsession-managers=[] -Dsnap=disabled"
   parallel:
     matrix:
       - CC: [gcc, clang]
@@ -322,7 +326,7 @@ build_release:
   extends:
     - .build_on_fedora
   variables:
-    MESON_OPTIONS: "-Dtest=enabled -Dbuildtype=release -Db_ndebug=true -Dsession-managers=[]"
+    MESON_OPTIONS: "-Dtest=enabled -Dbuildtype=release -Db_ndebug=true -Dsession-managers=[] -Dsnap=disabled"
   parallel:
     matrix:
       - CC: [gcc, clang]
@@ -336,7 +340,7 @@ build_session_managers:
     - ninja $NINJA_ARGS -C "$BUILD_DIR"
     - ninja $NINJA_ARGS -C "$BUILD_DIR" install
   variables:
-    MESON_OPTIONS: "-Dsession-managers=$SESSION_MANAGERS"
+    MESON_OPTIONS: "-Dsession-managers=$SESSION_MANAGERS -Dsnap=disabled"
   parallel:
     matrix:
       - SESSION_MANAGERS: ["[]", "wireplumber", "media-session", "media-session,wireplumber", "wireplumber,media-session" ]
@@ -352,7 +356,7 @@ build_meson_prerelease:
     - ninja $NINJA_ARGS -C "$BUILD_DIR"
     - ninja $NINJA_ARGS -C "$BUILD_DIR" install
   variables:
-    MESON_OPTIONS: "-Dsession-managers=wireplumber,media-session"
+    MESON_OPTIONS: "-Dsession-managers=wireplumber,media-session -Dsnap=disabled"
   allow_failure: true
 
 build_meson_exact_release:
@@ -369,7 +373,7 @@ build_meson_exact_release:
     - ninja $NINJA_ARGS -C "$BUILD_DIR"
     - ninja $NINJA_ARGS -C "$BUILD_DIR" install
   variables:
-    MESON_OPTIONS: "-Dsession-managers=[]"
+    MESON_OPTIONS: "-Dsession-managers=[] -Dsnap=disabled"
 
 valgrind:
   extends:
diff --git a/meson.build b/meson.build
index feca30ae1..5c797e835 100644
--- a/meson.build
+++ b/meson.build
@@ -397,6 +397,23 @@ lilv_lib = dependency('lilv-0', required: get_option('lv2'))
 summary({'lilv (for lv2 plugins)': lilv_lib.found()}, bool_yn: true)
 cdata.set('HAVE_LILV', lilv_lib.found())
 
+glib2_snap_dep = dependency('glib-2.0', required : get_option('snap'))
+gio2_snap_dep = dependency('gio-2.0', required : get_option('snap'))
+apparmor_snap_dep = dependency('libapparmor', required : get_option('snap'))
+if dependency('snapd-glib', required: false).found()
+  snap_dep = dependency('snapd-glib', required : get_option('snap'))
+else
+  snap_dep = dependency('snapd-glib-2', required : get_option('snap'))
+endif
+if snap_dep.found() and glib2_snap_dep.found() and gio2_snap_dep.found() and apparmor_snap_dep.found()
+  cdata.set('HAVE_SNAP', 1)
+  snap_deps = [glib2_snap_dep, gio2_snap_dep, snap_dep, apparmor_snap_dep]
+endif
+summary({'GLib-2.0 (Snap support)': glib2_snap_dep.found()}, bool_yn: true, section: 'Misc dependencies')
+summary({'Gio-2.0 (Snap support)': gio2_snap_dep.found()}, bool_yn: true, section: 'Misc dependencies')
+summary({'Apparmor (Snap support)': apparmor_snap_dep.found()}, bool_yn: true, section: 'Misc dependencies')
+summary({'Snapd-glib (Snap support)': snap_dep.found()}, bool_yn: true, section: 'Misc dependencies')
+
 check_functions = [
   ['gettid', '#include <unistd.h>', ['-D_GNU_SOURCE'], []],
   ['memfd_create', '#include <sys/mman.h>', ['-D_GNU_SOURCE'], []],
diff --git a/meson_options.txt b/meson_options.txt
index 99b057c52..7d2fe9aa0 100644
--- a/meson_options.txt
+++ b/meson_options.txt
@@ -285,3 +285,6 @@ option('compress-offload',
        description: 'Enable ALSA Compress-Offload support',
        type: 'feature',
        value: 'disabled')
+option('snap',
+       type : 'feature', value : 'auto',
+       description : 'Snap support is available.')
diff --git a/src/modules/meson.build b/src/modules/meson.build
index a3606a6cd..973124aac 100644
--- a/src/modules/meson.build
+++ b/src/modules/meson.build
@@ -265,6 +265,7 @@ pipewire_module_protocol_pulse_sources = [
   'module-protocol-pulse/reply.c',
   'module-protocol-pulse/sample.c',
   'module-protocol-pulse/sample-play.c',
+  'module-protocol-pulse/sandbox_access.c',
   'module-protocol-pulse/server.c',
   'module-protocol-pulse/stream.c',
   'module-protocol-pulse/utils.c',
@@ -295,6 +296,13 @@ pipewire_module_protocol_pulse_sources = [
   'module-protocol-pulse/modules/module-zeroconf-discover.c',
 ]
 
+if snap_dep.found() and glib2_snap_dep.found() and gio2_snap_dep.found() and apparmor_snap_dep.found()
+  pipewire_module_protocol_pulse_sources += [
+    'module-protocol-pulse/snap-policy.c',
+  ]
+  pipewire_module_protocol_pulse_deps += snap_deps
+endif
+
 if dbus_dep.found()
   pipewire_module_protocol_pulse_sources += [
     'module-protocol-pulse/dbus-name.c',
diff --git a/src/modules/module-protocol-pulse/pulse-server.c b/src/modules/module-protocol-pulse/pulse-server.c
index 41a814a42..c93720e57 100644
--- a/src/modules/module-protocol-pulse/pulse-server.c
+++ b/src/modules/module-protocol-pulse/pulse-server.c
@@ -75,6 +75,7 @@
 #include "stream.h"
 #include "utils.h"
 #include "volume.h"
+#include "sandbox_access.h"
 
 #define DEFAULT_MIN_REQ		"256/48000"
 #define DEFAULT_DEFAULT_REQ	"960/48000"
@@ -1604,6 +1605,9 @@ static int do_create_playback_stream(struct client *client, uint32_t command, ui
 	if (props == NULL)
 		goto error_errno;
 
+	if (!pw_sandbox_access_check_permissions(client, PW_SANDBOX_ACCESS_PLAYBACK))
+		goto error_no_permissions;
+
 	if (client->version < 13) {
 		if ((res = message_get(m,
 				TAG_STRING, &name,
@@ -1816,6 +1820,9 @@ error_protocol:
 error_no_formats:
 	res = -ENOTSUP;
 	goto error;
+error_no_permissions:
+	res = -EACCES;
+	goto error;
 error_invalid:
 	res = -EINVAL;
 	goto error;
@@ -1868,6 +1875,10 @@ static int do_create_record_stream(struct client *client, uint32_t command, uint
 	if (props == NULL)
 		goto error_errno;
 
+	// check permissions
+	if (!pw_sandbox_access_check_permissions(client, PW_SANDBOX_ACCESS_RECORD))
+		goto error_no_permissions;
+
 	if (client->version < 13) {
 		if (message_get(m,
 				TAG_STRING, &name,
@@ -2086,6 +2097,9 @@ error_protocol:
 error_no_formats:
 	res = -ENOTSUP;
 	goto error;
+error_no_permissions:
+	res = -EACCES;
+	goto error;
 error_invalid:
 	res = -EINVAL;
 	goto error;
diff --git a/src/modules/module-protocol-pulse/sandbox_access.c b/src/modules/module-protocol-pulse/sandbox_access.c
new file mode 100644
index 000000000..fd2161a8e
--- /dev/null
+++ b/src/modules/module-protocol-pulse/sandbox_access.c
@@ -0,0 +1,41 @@
+/* PipeWire Sandbox access manager
+ *
+ * Copyright © 2022 Sergio Costas (sergio.costas@canonical.com)
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice (including the next
+ * paragraph) shall be included in all copies or substantial portions of the
+ * Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ */
+
+#include <config.h>
+
+#include "sandbox_access.h"
+#include "snap-policy.h"
+
+bool pw_sandbox_access_check_permissions(struct client *client, pw_sandbox_access_t perms)
+{
+    pw_sandbox_access_t permissions;
+
+#ifdef HAVE_SNAP
+    permissions = snap_get_audio_permissions(client);
+    if ((permissions & PW_SANDBOX_ACCESS_NOT_A_SANDBOX) == 0)
+        return (perms & permissions) == perms;
+#endif // HAVE_SNAP
+    // add here checks for other sandbox technologies, like Flatpak
+    return true;
+}
diff --git a/src/modules/module-protocol-pulse/sandbox_access.h b/src/modules/module-protocol-pulse/sandbox_access.h
new file mode 100644
index 000000000..03dee60c5
--- /dev/null
+++ b/src/modules/module-protocol-pulse/sandbox_access.h
@@ -0,0 +1,42 @@
+/* PipeWire Sandbox access manager
+ *
+ * Copyright © 2022 Sergio Costas (sergio.costas@canonical.com)
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice (including the next
+ * paragraph) shall be included in all copies or substantial portions of the
+ * Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ */
+
+#ifndef _SANDBOX_ACCESS_H_
+#define _SANDBOX_ACCESS_H_
+
+#include "client.h"
+#include <stdbool.h>
+
+typedef enum _pw_sandbox_access {
+    PW_SANDBOX_ACCESS_NONE           = 0,
+    PW_SANDBOX_ACCESS_NOT_A_SANDBOX  = 1 << 0,
+    PW_SANDBOX_ACCESS_RECORD         = 1 << 1,
+    PW_SANDBOX_ACCESS_PLAYBACK       = 1 << 2,
+} pw_sandbox_access_t;
+
+#define PW_SANDBOX_ACCESS_ALL PW_SANDBOX_ACCESS_RECORD|PW_SANDBOX_ACCESS_PLAYBACK
+
+bool pw_sandbox_access_check_permissions(struct client *client, pw_sandbox_access_t permissions);
+
+#endif // _SANDBOX_ACCESS_H_
diff --git a/src/modules/module-protocol-pulse/snap-policy.c b/src/modules/module-protocol-pulse/snap-policy.c
new file mode 100644
index 000000000..bd700eb3f
--- /dev/null
+++ b/src/modules/module-protocol-pulse/snap-policy.c
@@ -0,0 +1,144 @@
+/***
+    This file is part of PipeWire.
+
+    Copyright 2022 Canonical Ltd.
+    Authors:
+      Sergio Costas <sergio.costas@canonical.com>
+
+    PipeWire is free software; you can redistribute it and/or modify
+    it under the terms of the GNU Lesser General Public License as published
+    by the Free Software Foundation; either version 2.1 of the License,
+    or (at your option) any later version.
+
+    PipeWire is distributed in the hope that it will be useful, but
+    WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+    General Public License for more details.
+
+    You should have received a copy of the GNU Lesser General Public License
+    along with PulseAudio; if not, see <http://www.gnu.org/licenses/>.
+***/
+
+#ifdef HAVE_CONFIG_H
+#include <config.h>
+#endif
+
+#include <glib.h>
+#include <snapd-glib/snapd-glib.h>
+#include <pipewire/pipewire.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include "client.h"
+#include <sys/apparmor.h>
+#include <errno.h>
+#include "snap-policy.h"
+#include <fcntl.h>
+
+#define SNAP_LABEL_PREFIX      "snap."
+#define SNAP_LABEL_PREFIX_SIZE 5
+
+pw_sandbox_access_t snap_get_audio_permissions(struct client *client)
+{
+    pid_t pid;
+    g_autofree gchar* aa_label = NULL;
+    g_autofree gchar* snap_id = NULL;
+    gchar* snap_confinement = NULL;
+    gchar *separator;
+    g_autoptr(SnapdClient) snapdclient = NULL;
+    g_autoptr(GPtrArray) plugs = NULL;
+    gboolean retv;
+    pw_sandbox_access_t permissions = PW_SANDBOX_ACCESS_NONE;
+    pw_sandbox_access_t add_permission = PW_SANDBOX_ACCESS_NONE;
+    SnapdPlug **plug = NULL;
+    GPtrArray *slots;
+    SnapdSlotRef **slot;
+
+    if (client == NULL) {
+        pw_log_warn("Called snap_get_audio_permissions with NULL parameter.");
+        return PW_SANDBOX_ACCESS_NONE;
+    }
+
+    // Get SNAP ID
+    if (pw_properties_fetch_int32(client->props, PW_KEY_APP_PROCESS_ID, &pid) != 0) {
+        pw_log_warn("snap_get_audio_permissions: failed to get 'application.process.id' property as an int32 value.");
+        return PW_SANDBOX_ACCESS_NONE;
+    }
+
+    if (aa_gettaskcon(pid, &aa_label, &snap_confinement) == -1) {
+        if (errno == EINVAL) {
+            // if apparmor isn't enabled, we can safely assume that there are no SNAPs in the system
+            return PW_SANDBOX_ACCESS_NOT_A_SANDBOX;
+        }
+        return PW_SANDBOX_ACCESS_NONE;
+    }
+    if (!g_str_has_prefix(aa_label, SNAP_LABEL_PREFIX)) {
+        // not a SNAP.
+        pw_log_info("snap_get_audio_permissions: not an snap.");
+        return PW_SANDBOX_ACCESS_NOT_A_SANDBOX;
+    }
+    // it's a "classic confinement" snap, so we give it full access
+    if (!g_strcmp0 (snap_confinement, "complain"))
+        return PW_SANDBOX_ACCESS_ALL;
+
+    snap_id = strdup(aa_label + SNAP_LABEL_PREFIX_SIZE);
+    separator = strchr(snap_id, '.');
+    if (separator != NULL)
+        *separator = 0;
+
+    snapdclient = snapd_client_new();
+    if (snapdclient == NULL) {
+        pw_log_warn("snap_get_audio_permissions: error creating SnapdClient object.");
+        return PW_SANDBOX_ACCESS_NONE;
+    }
+
+    retv = snapd_client_get_connections2_sync(snapdclient,
+                                              SNAPD_GET_CONNECTIONS_FLAGS_NONE,
+                                              snap_id,
+                                              NULL,
+                                              NULL,
+                                              NULL,
+                                              &plugs,
+                                              NULL,
+                                              NULL,
+                                              NULL);
+    if (retv == FALSE) {
+        pw_log_warn("Failed to get Snap connections\n");
+        return PW_SANDBOX_ACCESS_NONE;
+    }
+    if (plugs == NULL) {
+        pw_log_warn("Failed to get Snap connections\n");
+        return PW_SANDBOX_ACCESS_NONE;
+    }
+    if (plugs->pdata == NULL) {
+        pw_log_warn("Failed to get Snap connections\n");
+        return PW_SANDBOX_ACCESS_NONE;
+    }
+
+    plug = (SnapdPlug **)plugs->pdata;
+    for (guint p = 0; p < plugs->len; p++, plug++) {
+        const gchar *plug_name = snapd_plug_get_name(*plug);
+        if (!g_strcmp0("audio-record", plug_name)) {
+            add_permission = PW_SANDBOX_ACCESS_RECORD;
+        } else if (!g_strcmp0("audio-playback", plug_name)) {
+            add_permission = PW_SANDBOX_ACCESS_PLAYBACK;
+        } else if (!g_strcmp0("pulseaudio", plug_name)) {
+            add_permission = PW_SANDBOX_ACCESS_ALL;
+        } else {
+            continue;
+        }
+        slots = snapd_plug_get_connected_slots(*plug);
+        if (slots == NULL)
+            continue;
+        slot = (SnapdSlotRef**) slots->pdata;
+
+        for (guint q = 0; q < slots->len; q++, slot++) {
+            const gchar *slot_name = snapd_slot_ref_get_slot (*slot);
+            const gchar *snap_name = snapd_slot_ref_get_snap (*slot);
+            if ((0 == g_strcmp0 (snap_name, "snapd")) &&
+                (0 == g_strcmp0 (slot_name, plug_name)))
+                    permissions |= add_permission;
+        }
+        g_ptr_array_free(slots, TRUE);
+    }
+    return permissions;
+}
diff --git a/src/modules/module-protocol-pulse/snap-policy.h b/src/modules/module-protocol-pulse/snap-policy.h
new file mode 100644
index 000000000..25d8f0ebb
--- /dev/null
+++ b/src/modules/module-protocol-pulse/snap-policy.h
@@ -0,0 +1,29 @@
+/***
+    This file is part of PipeWire.
+
+    Copyright 2022 Canonical Ltd.
+    Authors:
+      Sergio Costas <sergio.costas@canonical.com>
+
+    PipeWire is free software; you can redistribute it and/or modify
+    it under the terms of the GNU Lesser General Public License as published
+    by the Free Software Foundation; either version 2.1 of the License,
+    or (at your option) any later version.
+
+    PipeWire is distributed in the hope that it will be useful, but
+    WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+    General Public License for more details.
+
+    You should have received a copy of the GNU Lesser General Public License
+    along with PulseAudio; if not, see <http://www.gnu.org/licenses/>.
+***/
+
+#ifndef _SNAP_POLICY_H_
+#define _SNAP_POLICY_H_
+
+#include "sandbox_access.h"
+
+pw_sandbox_access_t snap_get_audio_permissions(struct client *client);
+
+#endif // _SNAP_POLICY_H_
