diff --git a/.gitlab-ci.yml b/.gitlab-ci.yml
index 88ed826c8..e0760af26 100644
--- a/.gitlab-ci.yml
+++ b/.gitlab-ci.yml
@@ -83,6 +83,7 @@ include:
       debhelper-compat
       findutils
       git
+      libapparmor-dev
       libasound2-dev
       libavcodec-dev
       libavfilter-dev
@@ -94,6 +95,7 @@ include:
       libgstreamer-plugins-base1.0-dev
       libsbc-dev
       libsdl2-dev
+      libsnapd-glib-dev
       libudev-dev
       libva-dev
       libv4l-dev
@@ -228,7 +230,7 @@ build_on_ubuntu:
     - .build
   stage: build
   variables:
-    MESON_OPTIONS: "-Dsession-managers=[]"
+    MESON_OPTIONS: "-Dsession-managers=[] -Dsnap=enabled"
 
 .build_on_fedora:
   extends:
@@ -255,6 +257,7 @@ build_on_fedora:
         -Dsdl2=enabled
         -Dsndfile=enabled
         -Dsession-managers=[]
+        -Dsnap=disabled
   artifacts:
     name: pipewire-$CI_COMMIT_SHA
     when: always
@@ -270,7 +273,7 @@ build_on_alpine:
     - .build
   stage: build
   variables:
-    MESON_OPTIONS: "-Dsession-managers=[]"
+    MESON_OPTIONS: "-Dsession-managers=[] -Dsnap=disabled"
 
 # build with all auto() options enabled
 build_all:
@@ -287,6 +290,7 @@ build_all:
         -Droc=disabled
         -Dlibcamera=disabled
         -Dsession-managers=[]
+        -Dsnap=disabled
   parallel:
     matrix:
       - CC: [gcc, clang]
@@ -296,7 +300,7 @@ build_with_no_commandline_options:
   extends:
     - .build_on_fedora
   variables:
-    MESON_OPTIONS: "-Dsession-managers=[]"
+    MESON_OPTIONS: "-Dsession-managers=[] -Dsnap=disabled"
   parallel:
     matrix:
       - CC: [gcc, clang]
@@ -323,7 +327,7 @@ build_release:
   extends:
     - .build_on_fedora
   variables:
-    MESON_OPTIONS: "-Dtest=enabled -Dbuildtype=release -Db_ndebug=true -Dsession-managers=[]"
+    MESON_OPTIONS: "-Dtest=enabled -Dbuildtype=release -Db_ndebug=true -Dsession-managers=[] -Dsnap=disabled"
   parallel:
     matrix:
       - CC: [gcc, clang]
@@ -337,7 +341,7 @@ build_session_managers:
     - ninja $NINJA_ARGS -C "$BUILD_DIR"
     - ninja $NINJA_ARGS -C "$BUILD_DIR" install
   variables:
-    MESON_OPTIONS: "-Dsession-managers=$SESSION_MANAGERS"
+    MESON_OPTIONS: "-Dsession-managers=$SESSION_MANAGERS -Dsnap=disabled"
   parallel:
     matrix:
       - SESSION_MANAGERS: ["[]", "wireplumber", "media-session", "media-session,wireplumber", "wireplumber,media-session" ]
@@ -353,7 +357,7 @@ build_meson_prerelease:
     - ninja $NINJA_ARGS -C "$BUILD_DIR"
     - ninja $NINJA_ARGS -C "$BUILD_DIR" install
   variables:
-    MESON_OPTIONS: "-Dsession-managers=wireplumber,media-session"
+    MESON_OPTIONS: "-Dsession-managers=wireplumber,media-session -Dsnap=disabled"
   allow_failure: true
 
 build_meson_exact_release:
@@ -370,7 +374,7 @@ build_meson_exact_release:
     - ninja $NINJA_ARGS -C "$BUILD_DIR"
     - ninja $NINJA_ARGS -C "$BUILD_DIR" install
   variables:
-    MESON_OPTIONS: "-Dsession-managers=[]"
+    MESON_OPTIONS: "-Dsession-managers=[] -Dsnap=disabled"
 
 valgrind:
   extends:
diff --git a/meson.build b/meson.build
index e5f813ef7..7030a9a4f 100644
--- a/meson.build
+++ b/meson.build
@@ -399,6 +399,23 @@ lilv_lib = dependency('lilv-0', required: get_option('lv2'))
 summary({'lilv (for lv2 plugins)': lilv_lib.found()}, bool_yn: true)
 cdata.set('HAVE_LILV', lilv_lib.found())
 
+glib2_snap_dep = dependency('glib-2.0', required : get_option('snap'))
+gio2_snap_dep = dependency('gio-2.0', required : get_option('snap'))
+apparmor_snap_dep = dependency('libapparmor', required : get_option('snap'))
+if dependency('snapd-glib', required: false).found()
+  snap_dep = dependency('snapd-glib', required : get_option('snap'))
+else
+  snap_dep = dependency('snapd-glib-2', required : get_option('snap'))
+endif
+if snap_dep.found() and glib2_snap_dep.found() and gio2_snap_dep.found() and apparmor_snap_dep.found()
+  cdata.set('HAVE_SNAP', 1)
+  snap_deps = [glib2_snap_dep, gio2_snap_dep, snap_dep, apparmor_snap_dep]
+endif
+summary({'GLib-2.0 (Snap support)': glib2_snap_dep.found()}, bool_yn: true, section: 'Misc dependencies')
+summary({'Gio-2.0 (Snap support)': gio2_snap_dep.found()}, bool_yn: true, section: 'Misc dependencies')
+summary({'Apparmor (Snap support)': apparmor_snap_dep.found()}, bool_yn: true, section: 'Misc dependencies')
+summary({'Snapd-glib (Snap support)': snap_dep.found()}, bool_yn: true, section: 'Misc dependencies')
+
 check_functions = [
   ['gettid', '#include <unistd.h>', ['-D_GNU_SOURCE'], []],
   ['memfd_create', '#include <sys/mman.h>', ['-D_GNU_SOURCE'], []],
diff --git a/meson_options.txt b/meson_options.txt
index af72586b0..05e5287e0 100644
--- a/meson_options.txt
+++ b/meson_options.txt
@@ -318,3 +318,7 @@ option('rlimits-nice',
        min: -20,
        max: -1,
        value: -19)
+option('snap',
+       description : 'Snap support is available.',
+       type : 'feature',
+       value : 'auto')
diff --git a/src/modules/meson.build b/src/modules/meson.build
index 717fbf2a6..8e4a55908 100644
--- a/src/modules/meson.build
+++ b/src/modules/meson.build
@@ -264,6 +264,7 @@ pipewire_module_protocol_pulse_sources = [
   'module-protocol-pulse/reply.c',
   'module-protocol-pulse/sample.c',
   'module-protocol-pulse/sample-play.c',
+  'module-protocol-pulse/sandbox_access.c',
   'module-protocol-pulse/server.c',
   'module-protocol-pulse/stream.c',
   'module-protocol-pulse/utils.c',
@@ -294,6 +295,13 @@ pipewire_module_protocol_pulse_sources = [
   'module-protocol-pulse/modules/module-zeroconf-discover.c',
 ]
 
+if snap_dep.found() and glib2_snap_dep.found() and gio2_snap_dep.found() and apparmor_snap_dep.found()
+  pipewire_module_protocol_pulse_sources += [
+    'module-protocol-pulse/snap-policy.c',
+  ]
+  pipewire_module_protocol_pulse_deps += snap_deps
+endif
+
 if dbus_dep.found()
   pipewire_module_protocol_pulse_sources += [
     'module-protocol-pulse/dbus-name.c',
diff --git a/src/modules/module-protocol-pulse/pulse-server.c b/src/modules/module-protocol-pulse/pulse-server.c
index 0beda77a3..286413486 100644
--- a/src/modules/module-protocol-pulse/pulse-server.c
+++ b/src/modules/module-protocol-pulse/pulse-server.c
@@ -55,6 +55,7 @@
 #include "stream.h"
 #include "utils.h"
 #include "volume.h"
+#include "sandbox_access.h"
 
 #define DEFAULT_MIN_REQ		"256/48000"
 #define DEFAULT_DEFAULT_REQ	"960/48000"
@@ -1584,6 +1585,9 @@ static int do_create_playback_stream(struct client *client, uint32_t command, ui
 	if (props == NULL)
 		goto error_errno;
 
+	if (!pw_sandbox_access_check_permissions(client, PW_SANDBOX_ACCESS_PLAYBACK))
+		goto error_no_permissions;
+
 	if (client->version < 13) {
 		if ((res = message_get(m,
 				TAG_STRING, &name,
@@ -1785,6 +1789,9 @@ error_protocol:
 error_no_formats:
 	res = -ENOTSUP;
 	goto error;
+error_no_permissions:
+	res = -EACCES;
+	goto error;
 error_invalid:
 	res = -EINVAL;
 	goto error;
@@ -1837,6 +1844,9 @@ static int do_create_record_stream(struct client *client, uint32_t command, uint
 	if (props == NULL)
 		goto error_errno;
 
+	if (!pw_sandbox_access_check_permissions(client, PW_SANDBOX_ACCESS_RECORD))
+		goto error_no_permissions;
+
 	if (client->version < 13) {
 		if (message_get(m,
 				TAG_STRING, &name,
@@ -2044,6 +2054,9 @@ error_protocol:
 error_no_formats:
 	res = -ENOTSUP;
 	goto error;
+error_no_permissions:
+	res = -EACCES;
+	goto error;
 error_invalid:
 	res = -EINVAL;
 	goto error;
diff --git a/src/modules/module-protocol-pulse/sandbox_access.c b/src/modules/module-protocol-pulse/sandbox_access.c
new file mode 100644
index 000000000..9919051c5
--- /dev/null
+++ b/src/modules/module-protocol-pulse/sandbox_access.c
@@ -0,0 +1,47 @@
+/* PipeWire Sandbox access manager
+ *
+ * Copyright © 2022 Sergio Costas (sergio.costas@canonical.com)
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice (including the next
+ * paragraph) shall be included in all copies or substantial portions of the
+ * Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ */
+
+#include <config.h>
+#include <pipewire/pipewire.h>
+
+#include "sandbox_access.h"
+#include "snap-policy.h"
+
+bool pw_sandbox_access_check_permissions(struct client *client, pw_sandbox_access_t perms)
+{
+    pw_sandbox_access_t permissions;
+
+#ifdef HAVE_SNAP
+    permissions = PW_SANDBOX_ACCESS_NONE;
+    if (pw_properties_get(client->props, PW_KEY_SNAP_ID) != NULL) {
+        if (pw_properties_get_bool(client->props, PW_KEY_SNAP_PLAYBACK_ALLOWED, false))
+            permissions |= PW_SANDBOX_ACCESS_PLAYBACK;
+        if (pw_properties_get_bool(client->props, PW_KEY_SNAP_RECORD_ALLOWED, false))
+            permissions |= PW_SANDBOX_ACCESS_RECORD;
+        return (permissions & perms) == perms;
+    }
+#endif // HAVE_SNAP
+    // add here checks for other sandbox technologies, like Flatpak
+    return true;
+}
diff --git a/src/modules/module-protocol-pulse/sandbox_access.h b/src/modules/module-protocol-pulse/sandbox_access.h
new file mode 100644
index 000000000..0081b4a2e
--- /dev/null
+++ b/src/modules/module-protocol-pulse/sandbox_access.h
@@ -0,0 +1,58 @@
+/* PipeWire Sandbox access manager
+ *
+ * Copyright © 2022 Sergio Costas (sergio.costas@canonical.com)
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice (including the next
+ * paragraph) shall be included in all copies or substantial portions of the
+ * Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ */
+
+#ifndef _SANDBOX_ACCESS_H_
+#define _SANDBOX_ACCESS_H_
+
+#include "client.h"
+#include <stdbool.h>
+
+/**
+ * These are the permissions that a sandbox module can return.
+ * Currently, only Snap sandboxes are supported, so PW_SANDBOX_ACCESS_NOT_A_SANDBOX
+ * can seem redundant; but the idea is to be able to support more sandboxing
+ * technologies, like Flatpak. In this case, that value makes sense:
+ * if a sandbox module doesn't identify the process as belonging to a sandbox of
+ * their type, it will return PW_SANDBOX_ACCESS_NOT_A_SANDBOX, so the next sandbox
+ * module will be tried. If none of the sandbox modules detect the process as theirs,
+ * then PW_SANDBOX_ACCESS_ALL permissions will be granted, because it is presumed
+ * that it is a non-sandboxed process.
+ *
+ * To add other sandboxing systems, just modify *pw_sandbox_access_check_permissions()*
+ * function in *sandbox_access.c*, to include a call to get the permissions from
+ * your module.
+ */
+
+typedef enum _pw_sandbox_access {
+    PW_SANDBOX_ACCESS_NONE           = 0,
+    PW_SANDBOX_ACCESS_NOT_A_SANDBOX  = 1 << 0,
+    PW_SANDBOX_ACCESS_RECORD         = 1 << 1,
+    PW_SANDBOX_ACCESS_PLAYBACK       = 1 << 2,
+} pw_sandbox_access_t;
+
+#define PW_SANDBOX_ACCESS_ALL PW_SANDBOX_ACCESS_RECORD|PW_SANDBOX_ACCESS_PLAYBACK
+
+bool pw_sandbox_access_check_permissions(struct client *client, pw_sandbox_access_t required_permissions);
+
+#endif // _SANDBOX_ACCESS_H_
diff --git a/src/modules/module-protocol-pulse/server.c b/src/modules/module-protocol-pulse/server.c
index e01974eeb..9fa333fef 100644
--- a/src/modules/module-protocol-pulse/server.c
+++ b/src/modules/module-protocol-pulse/server.c
@@ -41,6 +41,9 @@
 #include "stream.h"
 #include "utils.h"
 #include "flatpak-utils.h"
+#ifdef HAVE_SNAP
+#include "snap-policy.h"
+#endif
 
 #define LISTEN_BACKLOG 32
 #define MAX_CLIENTS 64
@@ -405,6 +408,9 @@ on_connect(void *data, int fd, uint32_t mask)
 
 	if (server->addr.ss_family == AF_UNIX) {
 		char *app_id = NULL, *devices = NULL;
+#ifdef HAVE_SNAP
+		pw_sandbox_access_t snap_access;
+#endif
 
 #ifdef SO_PRIORITY
 		val = 6;
@@ -443,6 +449,15 @@ on_connect(void *data, int fd, uint32_t mask)
 			else
 				pw_properties_set(client->props, PW_KEY_MEDIA_CATEGORY, NULL);
 		}
+		// check SNAP permissions
+#ifdef HAVE_SNAP
+		snap_access = pw_snap_get_audio_permissions(client, client_fd, &app_id);
+		if ((snap_access & PW_SANDBOX_ACCESS_NOT_A_SANDBOX) == 0) {
+			pw_properties_set(client->props, PW_KEY_SNAP_ID, app_id);
+			pw_properties_set(client->props, PW_KEY_SNAP_PLAYBACK_ALLOWED, (snap_access & PW_SANDBOX_ACCESS_PLAYBACK) ? "1" : "0");
+			pw_properties_set(client->props, PW_KEY_SNAP_RECORD_ALLOWED, (snap_access & PW_SANDBOX_ACCESS_RECORD) ? "1" : "0");
+		}
+#endif
 		free(devices);
 		free(app_id);
 	}
diff --git a/src/modules/module-protocol-pulse/snap-policy.c b/src/modules/module-protocol-pulse/snap-policy.c
new file mode 100644
index 000000000..54fc4c279
--- /dev/null
+++ b/src/modules/module-protocol-pulse/snap-policy.c
@@ -0,0 +1,144 @@
+/***
+    This file is part of PipeWire.
+
+    Copyright 2022 Canonical Ltd.
+    Authors:
+      Sergio Costas <sergio.costas@canonical.com>
+
+    PipeWire is free software; you can redistribute it and/or modify
+    it under the terms of the GNU Lesser General Public License as published
+    by the Free Software Foundation; either version 2.1 of the License,
+    or (at your option) any later version.
+
+    PipeWire is distributed in the hope that it will be useful, but
+    WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+    General Public License for more details.
+
+    You should have received a copy of the GNU Lesser General Public License
+    along with PulseAudio; if not, see <http://www.gnu.org/licenses/>.
+***/
+
+#ifdef HAVE_CONFIG_H
+#include <config.h>
+#endif
+
+#include <glib.h>
+#include <snapd-glib/snapd-glib.h>
+#include <pipewire/pipewire.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include "client.h"
+#include <sys/apparmor.h>
+#include <errno.h>
+#include "snap-policy.h"
+#include <fcntl.h>
+
+#define SNAP_LABEL_PREFIX      "snap."
+
+static GHashTable *client_cache = NULL;
+
+pw_sandbox_access_t pw_snap_get_audio_permissions(struct client *client, int fd, char **app_id)
+{
+    pid_t pid;
+    g_autofree gchar* aa_label = NULL;
+    gchar* snap_id = NULL;
+    gchar* snap_confinement = NULL;
+    gchar *separator;
+    g_autoptr(SnapdClient) snapdclient = NULL;
+    g_autoptr(GPtrArray) plugs = NULL;
+    gboolean retv;
+    pw_sandbox_access_t permissions = PW_SANDBOX_ACCESS_NONE;
+    pw_sandbox_access_t add_permission = PW_SANDBOX_ACCESS_NONE;
+    SnapdPlug **plug = NULL;
+    GPtrArray *slots;
+    SnapdSlotRef **slot;
+
+    *app_id = strdup("unknown");
+    if (client == NULL) {
+        pw_log_warn("Called snap_get_audio_permissions with NULL parameter.");
+        return PW_SANDBOX_ACCESS_NONE;
+    }
+
+    if (aa_getpeercon(fd, &aa_label, &snap_confinement) == -1) {
+        if (errno == EINVAL) {
+            // if apparmor isn't enabled, we can safely assume that there are no SNAPs in the system
+            return PW_SANDBOX_ACCESS_NOT_A_SANDBOX;
+        }
+        pw_log_warn("snap_get_audio_permissions: failed to get the AppArmor info.");
+        return PW_SANDBOX_ACCESS_NONE;
+    }
+    if (!g_str_has_prefix(aa_label, SNAP_LABEL_PREFIX)) {
+        // not a SNAP.
+        pw_log_info("snap_get_audio_permissions: not an snap.");
+        return PW_SANDBOX_ACCESS_NOT_A_SANDBOX;
+    }
+
+    snap_id = strdup(aa_label + strlen(SNAP_LABEL_PREFIX));
+    separator = strchr(snap_id, '.');
+    if (separator != NULL)
+        *separator = 0;
+    *app_id = snap_id;
+
+    // it's a "classic" or a "devmode" confinement snap, so we give it full access
+    if (!g_strcmp0 (snap_confinement, "complain")) {
+        return PW_SANDBOX_ACCESS_ALL;
+    }
+
+    snapdclient = snapd_client_new();
+    if (snapdclient == NULL) {
+        pw_log_warn("snap_get_audio_permissions: error creating SnapdClient object.");
+        return PW_SANDBOX_ACCESS_NONE;
+    }
+
+    retv = snapd_client_get_connections2_sync(snapdclient,
+                                              SNAPD_GET_CONNECTIONS_FLAGS_NONE,
+                                              snap_id,
+                                              NULL,
+                                              NULL,
+                                              NULL,
+                                              &plugs,
+                                              NULL,
+                                              NULL,
+                                              NULL);
+    if (retv == FALSE) {
+        pw_log_warn("Failed to get Snap connections for snap %s\n", snap_id);
+        return PW_SANDBOX_ACCESS_NONE;
+    }
+    if (plugs == NULL) {
+        pw_log_warn("Failed to get Snap connections for snap %s\n", snap_id);
+        return PW_SANDBOX_ACCESS_NONE;
+    }
+    if (plugs->pdata == NULL) {
+        pw_log_warn("Failed to get Snap connections for snap %s\n", snap_id);
+        return PW_SANDBOX_ACCESS_NONE;
+    }
+
+    plug = (SnapdPlug **)plugs->pdata;
+    for (guint p = 0; p < plugs->len; p++, plug++) {
+        const gchar *plug_name = snapd_plug_get_name(*plug);
+        if (!g_strcmp0("audio-record", plug_name)) {
+            add_permission = PW_SANDBOX_ACCESS_RECORD;
+        } else if (!g_strcmp0("audio-playback", plug_name)) {
+            add_permission = PW_SANDBOX_ACCESS_PLAYBACK;
+        } else if (!g_strcmp0("pulseaudio", plug_name)) {
+            add_permission = PW_SANDBOX_ACCESS_ALL;
+        } else {
+            continue;
+        }
+        slots = snapd_plug_get_connected_slots(*plug);
+        if (slots == NULL)
+            continue;
+        slot = (SnapdSlotRef**) slots->pdata;
+
+        for (guint q = 0; q < slots->len; q++, slot++) {
+            const gchar *slot_name = snapd_slot_ref_get_slot (*slot);
+            const gchar *snap_name = snapd_slot_ref_get_snap (*slot);
+            if ((0 == g_strcmp0 (snap_name, "snapd")) &&
+                (0 == g_strcmp0 (slot_name, plug_name)))
+                    permissions |= add_permission;
+        }
+    }
+
+    return permissions;
+}
diff --git a/src/modules/module-protocol-pulse/snap-policy.h b/src/modules/module-protocol-pulse/snap-policy.h
new file mode 100644
index 000000000..78ba6a7e4
--- /dev/null
+++ b/src/modules/module-protocol-pulse/snap-policy.h
@@ -0,0 +1,42 @@
+/***
+    This file is part of PipeWire.
+
+    Copyright 2022 Canonical Ltd.
+    Authors:
+      Sergio Costas <sergio.costas@canonical.com>
+
+    PipeWire is free software; you can redistribute it and/or modify
+    it under the terms of the GNU Lesser General Public License as published
+    by the Free Software Foundation; either version 2.1 of the License,
+    or (at your option) any later version.
+
+    PipeWire is distributed in the hope that it will be useful, but
+    WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+    General Public License for more details.
+
+    You should have received a copy of the GNU Lesser General Public License
+    along with PulseAudio; if not, see <http://www.gnu.org/licenses/>.
+***/
+
+#ifndef _SNAP_POLICY_H_
+#define _SNAP_POLICY_H_
+
+#include "sandbox_access.h"
+
+/**
+ * These are the permissions that a sandbox module can return.
+ * The PW_SNAP_ACCESS_NOT_A_SANDBOX allows to differentiate between a sandbox
+ * with no audio permissions, and a program that is not a SNAP sandbox.
+ * This can happen if there is an error that precludes determining without doubt
+ * if the program is running or not in a SNAP sandbox.
+ */
+
+#define PW_KEY_SNAP_ID "pipewire.snap.id"
+#define PW_KEY_SNAP_PLAYBACK_ALLOWED "pipewire.snap.playback"
+#define PW_KEY_SNAP_RECORD_ALLOWED "pipewire.snap.record"
+
+pw_sandbox_access_t pw_snap_get_audio_permissions(struct client *client, int fd, char **app_id);
+bool pw_snap_check_audio_permissions(struct client *client, pw_sandbox_access_t required_permissions);
+
+#endif // _SNAP_POLICY_H_
