diff --git a/src/modules/module-protocol-pulse/snap-policy.c b/src/modules/module-protocol-pulse/snap-policy.c
index 2921f06de..23bccf293 100644
--- a/src/modules/module-protocol-pulse/snap-policy.c
+++ b/src/modules/module-protocol-pulse/snap-policy.c
@@ -193,7 +193,8 @@ pw_sandbox_access_t pw_snap_get_audio_permissions(struct client *client, int fd,
 			const gchar *slot_name = snapd_slot_ref_get_slot(*slot);
 			const gchar *snap_name = snapd_slot_ref_get_snap(*slot);
 			if ((g_str_equal(snap_name, "snapd") ||
-			     g_str_equal(snap_name, "core")) &&
+			     g_str_equal(snap_name, "core") ||
+			     g_str_equal(snap_name, "pipewire")) &&
 			     g_str_equal(slot_name, plug_name))
 				permissions |= add_permission;
 		}
